realtime-sanitizer/llvm-project
# Protect aligned_alloc and shared_mutex and tests behind ifdefs

This represents the proposed changes in https://github.com/realtime-sanitizer/radsan/issues/3

This change:

* Removes the hardcoded min version of 15
* Introduces a new cache variable called `SANITIZER_MIN_OSX_VERSION` that overrides this version
* Puts aligned_alloc and tests behind a protective block for OSX 10.15 and greater
* Puts shared_alloc and tests behind a protective block for OSX 10.12 and greater 

Please also note:
* Changed the helper function template `expectRealtimeDeath` to not use optional, as they were not introduced in an old enough OSX version. I opted for nullptr, as it provides this same functionality but with older compatibility.



To test this change:

1. Ensure that compiling and testing with OSX version 10.10 works:

```
cmake -G Ninja \
                                           -DCMAKE_BUILD_TYPE=Release \
                                           -DBUILD_SHARED_LIBS=ON \
                                           -DCOMPILER_RT_BUILD_SANITIZERS=ON \
                                           -DLLVM_ENABLE_PROJECTS="clang;compiler-rt" \
                                           -DLLVM_TARGETS_TO_BUILD=Native \
                                           -DCMAKE_C_FLAGS="" \
                                           -DCMAKE_CXX_FLAGS="" \
                                           -DSANITIZER_MIN_OSX_VERSION="10.10" \
                                           ../llvm/

> ninja TRadsan-arm64-Test && ./projects/compiler-rt/lib/radsan/tests/Radsan-arm64-Test | grep -E "PASSED|shared|aligned"
[1/1] Checking that just-built clang can find C++ headers...
[  PASSED  ] 58 tests.
```

Should compile and test with no errors, should NOT see any evidence of the shared_mutex, or aligned_alloc test running


2. Ensure "" with 10.12

```
cmake -G Ninja \
                                           -DCMAKE_BUILD_TYPE=Release \
                                           -DBUILD_SHARED_LIBS=ON \
                                           -DCOMPILER_RT_BUILD_SANITIZERS=ON \
                                           -DLLVM_ENABLE_PROJECTS="clang;compiler-rt" \
                                           -DLLVM_TARGETS_TO_BUILD=Native \
                                           -DCMAKE_C_FLAGS="" \
                                           -DCMAKE_CXX_FLAGS="" \
                                           -DSANITIZER_MIN_OSX_VERSION="10.12" \
                                           ../llvm/
> ninja TRadsan-arm64-Test && ./projects/compiler-rt/lib/radsan/tests/Radsan-arm64-Test | grep -E "PASSED|shared|aligned"
[1/1] Checking that just-built clang can find C++ headers...
[ RUN      ] TestRadsan.sharedLockingASharedMutexDiesWhenRealtime
[       OK ] TestRadsan.sharedLockingASharedMutexDiesWhenRealtime (79 ms)
[ RUN      ] TestRadsan.sharedUnlockingASharedMutexDiesWhenRealtime
[       OK ] TestRadsan.sharedUnlockingASharedMutexDiesWhenRealtime (78 ms)
[  PASSED  ] 62 tests.
```
Tests will pass
Aligned alloc will NOT appear
shared_mutex WILL appear

3. Ensure "" with 10.15

```
cmake -G Ninja \
                                           -DCMAKE_BUILD_TYPE=Release \
                                           -DBUILD_SHARED_LIBS=ON \
                                           -DCOMPILER_RT_BUILD_SANITIZERS=ON \
                                           -DLLVM_ENABLE_PROJECTS="clang;compiler-rt" \
                                           -DLLVM_TARGETS_TO_BUILD=Native \
                                           -DCMAKE_C_FLAGS="" \
                                           -DCMAKE_CXX_FLAGS="" \
                                           -DSANITIZER_MIN_OSX_VERSION="10.15" \
                                           ../llvm/

> ninja TRadsan-arm64-Test && ./projects/compiler-rt/lib/radsan/tests/Radsan-arm64-Test | grep -E "PASSED|shared|aligned"
[1/1] Checking that just-built clang can find C++ headers...
[ RUN      ] TestRadsan.sharedLockingASharedMutexDiesWhenRealtime
[       OK ] TestRadsan.sharedLockingASharedMutexDiesWhenRealtime (80 ms)
[ RUN      ] TestRadsan.sharedUnlockingASharedMutexDiesWhenRealtime
[       OK ] TestRadsan.sharedUnlockingASharedMutexDiesWhenRealtime (78 ms)
[ RUN      ] TestRadsanInterceptors.alignedAllocDiesWhenRealtime
[       OK ] TestRadsanInterceptors.alignedAllocDiesWhenRealtime (78 ms)
[  PASSED  ] 64 tests.
```

Tests pass
shared_mutex WILL appear
aligned_alloc WILL appear

4. Ensure compiling is OK on a non-OSX box

In linux, the aligned alloc test should appear and run
